### ISSUE TITLE
./bin/sp restart report {'cannot load',asn1rt_nif,get_files} and can not start 

### ISSUE THREAD

**Issue by mingchaoyan** [rebar/rebar#541]
##Environment

    Mac OS X 10.10.5
    Erlang/OTP: 17.5
    rebar 2.6.0 17 20150818_094957 git 2.6.0-16-g3239e74-dirty

It is an example of application from the book Programming Erlang. I have git push to [github][1]

Steps to reproduce the error

##build
```sh
$ git clone https://github.com/mingchaoyan/sellaprime.git
...
$ cd sellaprime
$ rebar prepare-deps
...
$ rebar compile
...
```

##release
```sh
$ mkdir rel
$ cd rel
$ rebar create-node nodeid=sp
...
```
edit reltool.config as follow
```erlang
%% -*- mode: erlang -*-
%% ex: ft=erlang
{sys, [
       {lib_dirs, []},
       {erts, [{mod_cond, derived}, {app_file, strip}]},
       {app_file, strip},
       {rel, "sp", "1.0.0",
        [
         kernel,
         stdlib,
         sasl,
         goldrush,
         lager,
         sp
        ]},
       {rel, "start_clean", "",
        [
         kernel,
         stdlib
        ]},
       {boot_rel, "sp"},
       {profile, embedded},
       {incl_cond, derived},
       {excl_archive_filters, [".*"]}, %% Do not archive built libs
       {excl_sys_filters, ["^bin/(?!start_clean.boot)",
                           "^erts.*/bin/(dialyzer|typer)",
                           "^erts.*/(doc|info|include|lib|man|src)"]},
       {excl_app_filters, ["\.gitignore"]},
       {app, goldrush, [{mod_cond, app}, {incl_cond, include}, {lib_dir, "../deps/goldrush"}]},
       {app, lager, [{mod_cond, app}, {incl_cond, include}, {lib_dir, "../deps/lager"}]},
       {app, sp, [{mod_cond, app}, {incl_cond, include}, {lib_dir, ".."}]}
      ]}.

{target_dir, "sp"}.

{overlay, [
           {mkdir, "log/sasl"},
           {copy, "files/erl", "\{\{erts_vsn\}\}/bin/erl"},
           {copy, "files/nodetool", "releases/\{\{rel_vsn\}\}/nodetool"},
           {copy, "sp/bin/start_clean.boot",
                  "\{\{erts_vsn\}\}/bin/start_clean.boot"},
           {copy, "files/sp", "bin/sp"},
           {copy, "files/sp.cmd", "bin/sp.cmd"},
           {copy, "files/start_erl.cmd", "bin/start_erl.cmd"},
           {copy, "files/sys.config", "releases/\{\{rel_vsn\}\}/sys.config"},
           {copy, "files/vm.args", "releases/\{\{rel_vsn\}\}/vm.args"}
          ]}.
```
then
```sh
$ rebar generate
...
```

##start
```sh
$ cd sp
$ ./bin/sp start
$ ./bin/sp getpid
...
$ ./bin/sp attach
```

start an another erl shell
```sh
erl -name 'other@127.0.0.1' -noinput -noshell -eval "rpc:call('sp@127.0.0.1', init, restart, [])" -setcookie sp
```
then the first erlang shell will report
````erlang
sp@127.0.0.1)1> {"init terminating in do_boot",{'cannot load',asn1rt_nif,get_files}}
```
How this happened? 

And why asn1rt will be released, actually I didn't config it in reltool.config


Thank you!

  [1]: https://github.com/mingchaoyan/sellaprime

**Comment by mingchaoyan** [comment]
It seems restart elang node by start_clean.boot

Why?  Is there any wrong config?

My rebar version is 2.6.0

**Comment by mingchaoyan** [comment]
I get the log {"init terminating in do_boot",{'cannot load',asn1rt_nif,get_files}} . 
How this happened? 
And why asn1rt will be released, actually I didn't config it in reltool.config

**Comment by mingchaoyan** [comment]
same to https://github.com/rebar/rebar/issues/45

**Comment by lrascao** [comment]
if you specify in reltool.config `{profile, development}` the restart will work, btw you can just attach to the node and do `init:restart()`to achieve the same result. 
I don't think the error conditions are exactly the same as in #45, by activating -loader_debug we can see that the error originates from loading asn1 and not crypto

**Comment by mingchaoyan** [comment]
@lrascao  what does profile here mean?  Where can I find the all options?

**Comment by lrascao** [comment]
Fred's book is a good resource on release handling: http://learnyousomeerlang.com/release-is-the-word
`{profile, development | embedded | standalone}
This option will act as a way to specify default *_filters (described below) based on your type of release. By default, development is used. That one will include more files from each app and ERTS blindly. The standalone profile will be more restrictive, and the embedded profile even more so, dropping more default ERTS applications and binaries.`

**Comment by mingchaoyan** [comment]
@lrascao I said the issue same to #45 because both can solve the problem by remove "-mode embedded". I just believe I had understand the difference between the embedded mode and interactive mode. The asn1 is not basically the modules needed. But now I check the ./bin/sp generate by rebar with {profile, development}, find the start mode is still embedded, and return to confused.   
What is the difference between  development  and embedded in this case? 
When to use development profile and embedded profile?
I just not to include asn1(only kernel,stdlib, sasl, goldrush,lager and sp), why this application want to start?
Thanks a lot!!!

**Comment by lrascao** [comment]
I finally tracked this down to a bug in OTP (the full explanation is at https://github.com/erlang/otp/pull/830 if curious), hopefully it will be included in a future release.
From what i could determine embedded mode does not allow you to do hot-code loading, this accounts for the manual entry that it's a much more restrictive mode of operation.
The ./bin/sp file is not generated by rebar, it is simply copied from rel/fles/sp when you issue the rebar generate command.

**Comment by lrascao** [comment]
ping @tuncer

**Comment by ferd** [comment]
Hot code loading can work with hot code loading, the difference is just that it won't be automated on need; each module needs to be explicitly loaded with `l(Mod)` or a call to the code server, iirc.

**Comment by tuncer** [comment]
@lrascao, thanks for investigating this. Commented in the pull request.

**Comment by lrascao** [comment]
PR https://github.com/erlang/otp/pull/830 has since been merged and released in OTP 18.2 (http://www.erlang.org/download/otp_src_18.2.readme, issue OTP-13115)
